src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingPasteCommandHandlercs.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export]
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.PasteTrackingPaste)]
+    [Order(After = PredefinedCommandHandlerNames.FormatDocument)]
+    [Order(Before = PredefinedCommandHandlerNames.Completion)]
+    internal partial class PasteTrackingPasteCommandHandler : IChainedCommandHandler<PasteCommandArgs>
+    {
+        public string DisplayName => EditorFeaturesResources.Paste_Tracking;
+
+        private readonly IPasteTrackingService _pasteTrackingService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        internal PasteTrackingPasteCommandHandler(IPasteTrackingService pasteTrackingService)
+        {
+            _pasteTrackingService = pasteTrackingService;
+        }
+
+        public VSCommanding.CommandState GetCommandState(PasteCommandArgs args, Func<VSCommanding.CommandState> nextCommandHandler)
+        {
+            return nextCommandHandler();
+        }
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            var document = args.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
+            }
+
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = TextSpan.FromBounds(snapshotSpan.Start, snapshotSpan.End);
```

Comment:
all this was to get this TextSpan... but all this will be is the position of the caret with zero-length afaict.

Reply:
Applying the post-paste snapshot to the tracking span creates a snapshot span that is the span of pasted text.

Reply:
yup.  ignore all this.  your explanation makes sense!